contracts/WitmonERC721.sol

```diff
@@ -0,0 +1,428 @@
+	{
+		return _state.creatures[_genotype];
+	}
+
+	function getCreatureStatus(bytes32 _genotype)
+		public view
+		virtual override
+		returns (Witmons.CreatureStatus)
+	{
+		Witmons.Creature storage _creature = _state.creatures[_genotype];
+		if (_creature.index == 0) {
+			return Witmons.CreatureStatus.Inexistent;
+		} else if (_creature.phenotype != bytes32(0)) {
+			return Witmons.CreatureStatus.Alive;
+		} else {
+			if (_creature.batchId != _state.lastBatch) {
+				return Witmons.CreatureStatus.Freezed;
+			}
+			Witmons.Status _tenderStatus = _state.status();
+			if (_tenderStatus == Witmons.Status.Hatching) {
+				return Witmons.CreatureStatus.Hatching;
+			} else {
+				return Witmons.CreatureStatus.Incubating;
+			}
+		}
+	}
+
+	function getGenotypes()
+		external view
+		override
+		returns (bytes32[] memory)
+	{
+		return _state.genotypes;
+	}
+
+	function getLastBatch()
+		external view
+		override
+		returns (uint256)
+	{
+		return _state.lastBatch;
+	}
+
+	function getStats()
+		public view
+		override
+		returns (
+			uint256 _totalSuppy,
+			uint256 _totalBatches,
+			uint256 _totalEggs,
+			uint256 _totalHatches,
+			uint256 _totalGenotypes
+		)
+	{
+		return (
+			_state.tokenIds.current(),
+			_state.totalBatches,
+			_state.totalEggs,
+			_state.totalHatches,
+			_state.genotypes.length
+		);
+	}
+
+	function getStatus()
+		public view
+		override
+		returns (Witmons.Status)
+	{
+		return _state.status();
+	}
+
+	function getTokenGenotype(uint256 _tokenId)
+		public view
+		override
+		tokenExists(_tokenId)
+		returns (bytes32)
+	{
+		return _state.tokenGenotypes[_tokenId];
+	}
+
+	function getTokenPicture(uint256 _tokenId)
+		external view
+		override
+		returns (string memory)
+	{
+		bytes32 _genotype = getTokenGenotype(_tokenId);
+		Witmons.Creature memory _creature = _state.creatures[_genotype];
+		Witmons.Batch storage _batch = _state.batches[_creature.batchId];
+		address _decorator = _batch.decorator;
+		return (_decorator != address(0))
+			? IWitmonDecorator(_decorator).picture(_genotype, _creature)
+			: ''
+		;
+	}
+
+
+	// ------------------------------------------------------------------------
+	// --- INTERNAL METHODS ---------------------------------------------------
+	// ------------------------------------------------------------------------
+
+	function _getLastBatch()
+		internal view
+		returns (
+			uint256 _batchId,
+			Witmons.Batch storage _batch
+		)
+	{
+		_batchId = _state.lastBatch;
+		_batch = _state.batches[_batchId];
+	}
+
+
+	// ------------------------------------------------------------------------
+	// --- PRIVATE METHODS ----------------------------------------------------
+	// ------------------------------------------------------------------------
+
+	function _bytesToBytes32(bytes memory _bb)
+		private pure
+		returns (bytes32 _r)
+	{
+		uint _len = _bb.length > 32 ? 32 : _bb.length;
+		for (uint _i = 0; _i < _len; _i ++) {
+			_r |= bytes32(_bb[_i] & 0xff) >> (_i * 8);
```

Comment:
This doesn't look right, `0xff >> 8` evaluates to 0.

Reply:
0xff00000000000000000000000000000000000000000000000000000000000000 does, though.

Reply:
I see, so bytes32(x) converts it to 0xff00000..., I expected it to be 0x0000...00ff.